### ISSUE TITLE
EF trying to Insert into Identity field

### ISSUE THREAD

**Issue by Dethorhyne** [aspnet/EntityFramework#6426]
### Steps to reproduce
1. Make a new solution, add NumericId to the ApplicationUser class and make it look like this:
```
public class ApplicationUser : IdentityUser
{
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public virtual int NumericId { get; set; }
}
```
2. Execute UserManager.UpdateAsync(user);
3. Witness the error

### The issue
For some reason EF is attempting to insert value into an Identity column which isn't allowed and throws an exception. Rest of the columns do NOT get updated.

However,
this bug also accured when I tried to register, but in that case, database wise, everything was Inserted, new user was registered and a proper NumericId was assigned to the user even though the front-end threw the exception. Few builds after, even though nothing has changed, registering stopped throwing the exception and functions properly.

```
Exception message:

SqlException: Cannot update identity column 'NumericId'.
MoveNext

DbUpdateException: An error occurred while updating the entries. See the inner exception for details.
MoveNext

Stack trace:
System.Data.SqlClient.SqlException: Cannot update identity column 'NumericId'.
   at System.Data.SqlClient.SqlCommand.<>c.<ExecuteDbDataReaderAsync>b__107_0(Task`1 result)
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.<ExecuteAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.<ExecuteAsync>d__32.MoveNext()
ClientConnectionId:5be684a0-2aee-4a9f-ac8b-9d4bcea49b47
Error Number:8102,State:1,Class:16
```

### Further technical details

EF Core version: 1.0.0
Operating system: Windows 7 Professional
Visual Studio version: VS 2015 Enterprise edition

Other details about my project setup:
None.

**Comment by AndriySvyryd** [comment]
@Dethorhyne This is a bug in how we handle the data annotation. You can work around this by using Fluent API in `OnModelCreating`:
```C#
modelBuilder
    .Entity<ApplicationUser>()
    .Property(u => u.NumericId)
    .UseSqlServerIdentityColumn();
```

**Comment by brentdavid2008** [comment]
i am still getting this with 1.1.1 and VS 2017.  The workaround using the fluent API did not seem to change anything.

![image](https://cloud.githubusercontent.com/assets/7902547/25316553/f723992a-282e-11e7-843e-33847700e25a.png)

![image](https://cloud.githubusercontent.com/assets/7902547/25316557/042b95f0-282f-11e7-9b51-a6e750a44cd4.png)

![image](https://cloud.githubusercontent.com/assets/7902547/25316562/10dfbd76-282f-11e7-827f-9158872b329f.png)

web csproj packages: 

![image](https://cloud.githubusercontent.com/assets/7902547/25316575/3faeed3e-282f-11e7-85d5-59fedc6d150c.png)

Suggestions?

**Comment by ajcvickers** [comment]
@brentdavid2008 Can you please open a new issue including a complete code listing or project that reproduces what you are seeing.

**Comment by zahik** [comment]
+1.
Same problem.

**Comment by mkiani3000** [comment]
Use Metadata property like this : 

**Before Entity Frmework core 2.0 :**
```
 builder.Property(p => p.Id)
                .UseSqlServerIdentityColumn();
 builder.Property(p => p.Id)
             .Metadata.IsReadOnlyAfterSave = true;
```

**for Entity framework core 2.0 , The '`IsReadOnlyAfterSave` property ' is deprecated.Use following :** 
          ```
  builder.Property(p => p.Id)
                .UseSqlServerIdentityColumn();
            builder.Property(p => p.Id)
                .Metadata.AfterSaveBehavior = PropertySaveBehavior.Ignore;
```

**Comment by Gimly** [comment]
@mkiani3000 I tried to do exactly as you explained on EF Core 2.0 and I'm getting this error message when running `Add-Migration`:

> The property 'Id' on entity type 'Scenario' must be marked as read-only after it has been saved because it is part of a key. Key properties are always read-only once an entity has been saved for the first time.

Here's the code in the `OnModelCreating`
```C#
var scenarioBuilder = modelBuilder.Entity<Person>();

scenarioBuilder.Property(p => p.Id).UseSqlServerIdentityColumn();
scenarioBuilder.Property(p => p.Id).Metadata.AfterSaveBehavior = PropertySaveBehavior.Ignore;

```

**Comment by ajcvickers** [comment]
@Gimly Keys must be marked as PropertySaveBehavior.Throw--key values cannot be changed after the entity has been saved. This is the default, so nothing should need to be set.

**Comment by LoopsLu** [comment]
I had this issue too!!

Try the method below. I use it to fix my problem.

Please add these code in your Context.cs
`protected override void OnModelCreating(ModelBuilder modelBuilder)
{
   base.OnModelCreating(modelBuilder);

   modelBuilder.Entity<User>()
   .HasAlternateKey(x => x.NumericId ).HasName("IX_NumericId");
}

**Comment by Vaishnavi6121995** [comment]
Hello @ajcvickers I m getting this type of error 

{"message":"An error occurred while updating the entries. See the inner exception for details.","detail":"   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.ExecuteAsync(IRelationalConnection connection, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.ExecuteAsync(DbContext _, ValueTuple`2 parameters, CancellationToken cancellationToken)\r\n   at Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.NpgsqlExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken) in C:\\projects\\npgsql-entityframeworkcore-postgresql\\src\\EFCore.PG\\Storage\\Internal\\NpgsqlExecutionStrategy.cs:line 49\r\n   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(IReadOnlyList`1 entriesToSave, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)\r\n   at Microsoft.EntityFrameworkCore.DbContext.SaveChangesAsync(Boolean acceptAllChangesOnSuccess, CancellationToken cancellationToken)\r\n   at SmartAssistant.LinkMe.Repository.RepositoryBase`1.SaveAsync() in D:\\Projects\\ClientProjects\\Recruitment_Agency_Project\\SmartAssistance\\smartassistantsa-smartassistant.whosfit-5aad80240237\\server\\SmartAssistant.LinkMe.Repository\\RepositoryBase.cs:line 48\r\n   at SmartAssistant.LinkMe.Repository.EmployerRepository.CreateEmployerAsync(EmployerDBM EmployerDBM) in D:\\Projects\\ClientProjects\\Recruitment_Agency_Project\\SmartAssistance\\smartassistantsa-smartassistant.whosfit-5aad80240237\\server\\SmartAssistant.LinkMe.Repository\\EmployerRepository.cs:line 45\r\n   at SmartAssistant.LinkMe.Api.Controllers.EmployerController.CreateEmployee(EmployerDBM Employeer) in D:\\Projects\\ClientProjects\\Recruitment_Agency_Project\\SmartAssistance\\smartassistantsa-smartassistant.whosfit-5aad80240237\\server\\SmartAssistant.LinkMe.Api\\Controllers\\EmployerController.cs:line 69\r\n   at Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor.TaskOfIActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)\r\n   at System.Threading.Tasks.ValueTask`1.get_Result()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeActionMethodAsync()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeNextActionFilterAsync()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ActionExecutedContext context)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.InvokeInnerFilterAsync()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeFilterPipelineAsync()\r\n   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeAsync()\r\n   at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)\r\n   at Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(HttpContext httpContext)\r\n   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.Invoke(HttpContext context)"}

**Comment by ajcvickers** [comment]
@Vaishnavi6121995 Please create a new issue and include a small, runnable project/solution or complete code listing that demonstrates the exception you are seeing.

**Comment by simeyla** [comment]
I'm guessing this is now 

     scenarioBuilder.Property(p => p.Id).UseIdentityColumn();    // in SqlServerPropertyBuilderExtensions
